scipy/special/_logsumexp.py

```diff
@@ -104,19 +104,43 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
     elif not np.isfinite(a_max):
         a_max = 0
 
+    tmp0 = (a == a_max) # compute largest separately
+
+
     if b is not None:
         b = np.asarray(b)
-        tmp = b * np.exp(a - a_max)
+        tmp = b * np.exp(a - a_max)  * (~tmp0)
+        tmp0 = b*(tmp0 + .0)
     else:
-        tmp = np.exp(a - a_max)
+        tmp = np.exp(a - a_max)  * (~tmp0)
```

Comment:
So this pulls out the values where a == a_max to be able to use log1p below. Needs a comment (something along the lines of what Warren wrote in the OP). One option is to write a formula in a comment.